File: plugin/src/main/groovy/com/therouter/plugin/RouteItem.java
package com.therouter.plugin;

import java.util.HashMap;
import java.util.Objects;

public class RouteItem implements Comparable<RouteItem> {
    String path = "";
    String className = "";
    String action = "";
    String description = "";
    HashMap<String, String> params = new HashMap<>();

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof RouteItem)) return false;
        RouteItem routeItem = (RouteItem) o;
        return Objects.equals(path, routeItem.path) &&
                Objects.equals(className, routeItem.className);
    }

    @Override
    public int hashCode() {
        return Objects.hash(path, className, action, description, params);
    }

    @Override
    public int compareTo(RouteItem routeItem) {
        if (routeItem == null || routeItem.className == null || className == null) {
            return 0;
        }
        if (routeItem.className.compareTo(className) == 0) {
            if (routeItem.path == null) {
                throw new RuntimeException("TheRouter " + routeItem.className + "'s path is Null");
            }
            if (path == null) {
                throw new RuntimeException("TheRouter " + className + "'s path is Null");
            }
            return routeItem.path.compareTo(path);
        } else {
            return routeItem.className.compareTo(className);
        }
    }

    @Override
    public String toString() {
        return "{" +
                "path='" + path + '\'' +
                ", className='" + className + '\'' +
                ", action='" + action + '\'' +
                ", description='" + description + '\'' +
                ", params=" + params +
                '}';
    }
}

File: apt/src/main/java/com/therouter/apt/RouteItem.kt
package com.therouter.apt

import java.util.*

class RouteItem : Comparable<RouteItem?> {

    var path = ""

    var className: String? = ""

    var action = ""

    var description = ""

    var params = HashMap<String, String>()
    override fun compareTo(routeItem: RouteItem?): Int {
        return if (routeItem?.className == null || className == null) {
            0
        } else routeItem.className!!.compareTo(
            className!!
        )
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is RouteItem) return false
        return path == o.path &&
                className == o.className &&
                action == o.action &&
                description == o.description &&
                params == o.params
    }

    override fun hashCode(): Int {
        return Objects.hash(path, className, action, description, params)
    }
}